src/main/java/htsjdk/samtools/cram/structure/ContainerIO.java

```diff
@@ -131,26 +131,28 @@ public static int writeContainer(final Version version, final Container containe
                 if (isFileHeaderContainer) {
                     final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
                     firstBlock.write(version.major, byteArrayOutputStream);
-                    container.containerByteSize = byteArrayOutputStream.size();
+                    container.containerBlocksByteSize = byteArrayOutputStream.size();
 
                     final int containerHeaderByteSize = ContainerHeaderIO.writeContainerHeader(version.major, container, outputStream);
                     try {
-                        outputStream.write(byteArrayOutputStream.toByteArray(), 0, byteArrayOutputStream.size());
+                        outputStream.write(byteArrayOutputStream.toByteArray(), 0, container.containerBlocksByteSize);
                     } catch (final IOException e) {
                         throw new RuntimeIOException(e);
                     }
-                    return containerHeaderByteSize + byteArrayOutputStream.size();
+                    return containerHeaderByteSize + container.containerBlocksByteSize;
                 }
             }
         }
 
+        // use this BAOS for two purposes: writing out and counting bytes for landmarks/containerBlocksByteSize
         final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
-
         container.compressionHeader.write(version, byteArrayOutputStream);
         container.blockCount = 1;
 
         final List<Integer> landmarks = new ArrayList<>();
         for (final Slice slice : container.getSlices()) {
+            // landmark 0 = byte length of the compression header
+            // landmarks after 0 = byte length of the compression header plus all slices before this one
```

Comment:
It's worse than that: I don't think the embedded ref block should increment the count because that actually duplicates one of the external blocks.  But I'm not confident in that.

I'll add comments about what I think/know about these.